File: arduino/interpreter.c
#include "interpreter.h"

uint32_t executed = 0;
uint16_t exePointer = 0;
uint16_t stackPointer = 0;
bool (*callHandler)(uint8_t);

// optimizations:
// - push <addr:2> -> pushb <addr:1>
// - pop <addr:2> -> popb <addr:1>
// - instructions like add8

int run()
{
    while (1)
    {
        executed++;
        switch (mem[exePointer++])
        {
        case Noop:
            continue;
        case Push:
        {
            unsigned short addr = mem[exePointer++] | (mem[exePointer++] << 8);
            stackPointer -= 4;
            *(int32_t *)(mem + stackPointer) = *(int32_t *)(mem + addr);
#ifdef DEBUG
            printf("push %d\n", addr);
#endif
            continue;
        }
        case Pop:
        {
            unsigned short addr = mem[exePointer++] | (mem[exePointer++] << 8);
            *(int32_t *)(mem + addr) = *(int32_t *)(mem + stackPointer);
#ifdef DEBUG
            printf("pop %d\n", addr);
#endif
            stackPointer += 4;
            continue;
        }
        case PushConst8:
        {
            uint8_t val = mem[exePointer++];
            stackPointer -= 4;
            // mem[stackPointer] = val;
            *(int32_t *)(mem + stackPointer) = val;
#ifdef DEBUG
            printf("pushconst8 %d\n", val);
#endif
            continue;
        }
        case PushConst16:
        {
            uint16_t val = (uint16_t)mem[exePointer++] | (uint16_t)(mem[exePointer++] << 8);
            stackPointer -= 4;
            *(int32_t *)(mem + stackPointer) = val;
#ifdef DEBUG
            printf("pushconst16 %d\n", val);
#endif
            continue;
        }
        case PushConst32:
        {
            int32_t val = (int32_t)mem[exePointer++] | (int32_t)(mem[exePointer++] << 8) | (int32_t)(mem[exePointer++] << 16) | (int32_t)(mem[exePointer++] << 24);
            stackPointer -= 4;
            *(int32_t *)(mem + stackPointer) = val;
#ifdef DEBUG
            printf("pushconst32 %d\n", val);
#endif
            continue;
        }
        case Dup:
        {
            stackPointer -= 4;
            *(int32_t *)(mem + stackPointer) = *(int32_t *)(mem + stackPointer + 4);
#ifdef DEBUG
            printf("dup\n");
#endif
            continue;
        }
        case Swap:
        {
            int32_t temp = *(int32_t *)(mem + stackPointer + 4);
            *(int32_t *)(mem + stackPointer + 4) = *(int32_t *)(mem + stackPointer);
            *(int32_t *)(mem + stackPointer) = temp;
#ifdef DEBUG
            printf("swap\n");
#endif
            continue;
        }
        case Pop8:
        {
            unsigned short addr = mem[exePointer++] | (mem[exePointer++] << 8);
            mem[addr] = mem[stackPointer];
#ifdef DEBUG
            printf("pop8 %d\n", addr);
#endif
            stackPointer += 4;
            continue;
        }
        case Push8:
        {
            unsigned short addr = mem[exePointer++] | (mem[exePointer++] << 8);
            stackPointer -= 4;
            *(int32_t *)(mem + stackPointer) = mem[addr];
#ifdef DEBUG
            printf("push8 %d\n", addr);
#endif
            continue;
        }
        case Consume:
        {
            stackPointer += 4;
#ifdef DEBUG
            printf("consume\n");
#endif
            continue;
        }
        case 0x0B:
        case 0x0C:
        case 0x0D:
        case 0x0E:
            return EINVOP;

        case Halt:
        {
#ifdef DEBUG
            printf("halt (stackPointer at %x)\n", stackPointer);
#endif
            return 0;
        }
        case Add:
        {
            int32_t val = *(int32_t *)(mem + stackPointer);
            stackPointer += 4;
            *(int32_t *)(mem + stackPointer) += val;
#ifdef DEBUG
            printf("add\n");
#endif
            continue;
        }
        case Sub:
        {
            int32_t val = *(int32_t *)(mem + stackPointer);
            stackPointer += 4;
            *(int32_t *)(mem + stackPointer) -= val;
#ifdef DEBUG
            printf("sub\n");
#endif
            continue;
        }
        case Mul:
        {

            int32_t val = *(int32_t *)(mem + stackPointer);
            stackPointer += 4;
            *(int32_t *)(mem + stackPointer) *= val;
#ifdef DEBUG
            printf("mul\n");
#endif
            continue;
        }
        case Div:
        {

            int32_t val = *(int32_t *)(mem + stackPointer);
            stackPointer += 4;
            *(int32_t *)(mem + stackPointer) /= val;
#ifdef DEBUG
            printf("div\n");
#endif
            continue;
        }
        case Mod:
        {

            int32_t val = *(int32_t *)(mem + stackPointer);
            stackPointer += 4;
            *(int32_t *)(mem + stackPointer) %= val;
#ifdef DEBUG
            printf("mod\n");
#endif
            continue;
        }
        case Inv:
        {

            *(int32_t *)(mem + stackPointer) = -(*(int32_t *)(mem + stackPointer));
#ifdef DEBUG
            printf("inv\n");
#endif
            continue;
        }
        case Abs:
        {
            int32_t val = *(int32_t *)(mem + stackPointer);
            *(int32_t *)(mem + stackPointer) = val < 0 ? -val : val;
#ifdef DEBUG
            printf("abs\n");
#endif
            continue;
        }
        case Add8:
        {
            char val = mem[exePointer++];
            *(int32_t *)(mem + stackPointer) += val;
#ifdef DEBUG
            printf("add8 %d\n", val);
#endif
            continue;
        }
        case 0x18:
        case 0x19:
        case 0x1A:
        case 0x1B:
        case 0x1C:
        case 0x1D:
        case 0x1E:
        case 0x1F:
            return EINVOP;

        case Jrnz:
        {
            unsigned char rel = mem[exePointer++];
#ifdef DEBUG
            printf("jrnz %d\n", rel);
#endif
            if (*(int32_t *)(mem + stackPointer))
            {
                exePointer += rel;
            }
            stackPointer += 4;
            continue;
        }

        case Jrz:
        {
            unsigned char rel = mem[exePointer++];
#ifdef DEBUG
            printf("jrz %d\n", rel);
#endif
            if (!*(int32_t *)(mem + stackPointer))
            {
                exePointer += rel;
            }
            stackPointer += 4;
            continue;
        }
        case Jr:
        {
            unsigned char rel = mem[exePointer++];
#ifdef DEBUG
            printf("jr %d\n", rel);
#endif
            exePointer += rel;
            continue;
        }
        case Call:
        {
            unsigned char id = mem[exePointer++];
#ifdef DEBUG
            printf("call %d\n", id);
#endif
            if (!callHandler(id))
                return EINVCALL;
            continue;
        }

        case 0x24:
        case 0x25:
        case 0x26:
        case 0x27:
        case 0x28:
        case 0x29:
        case 0x2A:
        case 0x2B:
        case 0x2C:
        case 0x2D:
        case 0x2E:
        case 0x2F:
            return EINVOP;

        case Eq:
        {
            int32_t val = *(int32_t *)(mem + stackPointer);
            stackPointer += 4;
            *(int32_t *)(mem + stackPointer) = *(int32_t *)(mem + stackPointer) == val;
#ifdef DEBUG
            printf("eq\n");
#endif
            continue;
        }
        case Neq:
        {
            int32_t val = *(int32_t *)(mem + stackPointer);
            stackPointer += 4;
            *(int32_t *)(mem + stackPointer) = *(int32_t *)(mem + stackPointer) != val;
#ifdef DEBUG
            printf("neq\n");
#endif
            continue;
        }
        case Lt:
        {
            int32_t val = *(int32_t *)(mem + stackPointer);
            stackPointer += 4;
            *(int32_t *)(mem + stackPointer) = *(int32_t *)(mem + stackPointer) < val;
#ifdef DEBUG
            printf("lt\n");
#endif
            continue;
        }
        case Bt:
        {
            int32_t val = *(int32_t *)(mem + stackPointer);
            stackPointer += 4;
            *(int32_t *)(mem + stackPointer) = *(int32_t *)(mem + stackPointer) > val;
#ifdef DEBUG
            printf("bt\n");
#endif
            continue;
        }
        case Lte:
        {
            int32_t val = *(int32_t *)(mem + stackPointer);
            stackPointer += 4;
            *(int32_t *)(mem + stackPointer) = *(int32_t *)(mem + stackPointer) <= val;
#ifdef DEBUG
            printf("lte\n");
#endif
            continue;
        }
        case Bte:
        {
            int32_t val = *(int32_t *)(mem + stackPointer);
            stackPointer += 4;
            *(int32_t *)(mem + stackPointer) = *(int32_t *)(mem + stackPointer) >= val;
#ifdef DEBUG
            printf("bte\n");
#endif
            continue;
        }

        case 0x36:
        case 0x37:
        case 0x38:
        case 0x39:
        case 0x3A:
        case 0x3B:
        case 0x3C:
        case 0x3D:
        case 0x3E:
        case 0x3F:
            return EINVOP;

        case Sin:
        {
            // *(int32_t *)(mem + stackPointer) = sin(*(int32_t *)(mem + stackPointer) / 1000.0f) * 1000;
            *(int32_t *)(mem + stackPointer) = fastSin(mem[stackPointer]);
            continue;
        }
        case Cos:
        {
            // *(int32_t *)(mem + stackPointer) = cos(*(int32_t *)(mem + stackPointer) / 1000.0f) * 1000;
            *(int32_t *)(mem + stackPointer) = fastCos(mem[stackPointer]);
            continue;
        }

        default:
#ifdef DEBUG
            printf("unknown opcode %d\n", mem[exePointer - 1]);
#endif
            return EINVOP;
        }
    }
    return 0;
}

File: compiler/src/interpreter.ts
import { OpCode } from "./target";
import debug from "debug";

const info = debug("rgb:interpreter");

export class Interpreter {
    readonly memory: Uint8Array;
    stackPointer: number;
    exePointer: number;
    debug = false;

    constructor(memory: Uint8Array, entryPoint: number, memorySize = 2 ** 16) {
        this.memory = new Uint8Array(memorySize);
        for (let i = 0; i < memory.length; i++) {
            this.memory[i] = memory[i];
        }
        this.exePointer = entryPoint;
        this.stackPointer = memorySize;
    }

    writeInt(pos: number, value: number) {
        this.memory[pos] = value & 0x000000ff;
        this.memory[pos + 1] = value & 0x0000ff00;
        this.memory[pos + 2] = value & 0x00ff0000;
        this.memory[pos + 3] = value & 0xff000000;
    }

    readInt(pos: number): number {
        return this.memory[pos] | (this.memory[pos + 1] << 8) | (this.memory[pos + 2] << 16) | (this.memory[pos + 3] << 24);
    }

    executeNext() {
        let op: OpCode = this.memory[this.exePointer++];
        switch (op) {
            case OpCode.Noop:
                break;
            case OpCode.Push: {
                let addr = this.memory[this.exePointer++] | (this.memory[this.exePointer++] << 8);
                if (this.debug) info("push", addr);
                this.stackPointer -= 4;
                this.writeInt(this.stackPointer, this.readInt(addr));
                break;
            }
            case OpCode.Pop: {
                let addr = this.memory[this.exePointer++] | (this.memory[this.exePointer++] << 8);
                if (this.debug) info("pop", addr);
                this.writeInt(addr, this.readInt(this.stackPointer));
                this.stackPointer += 4;
                break;
            }
            case OpCode.PushConst8: {
                let value = this.memory[this.exePointer++];
                if (this.debug) info("pushconst8", value);
                this.stackPointer -= 4;
                this.writeInt(this.stackPointer, value);
                break;
            }
            case OpCode.PushConst16: {
                let value = this.memory[this.exePointer++] | (this.memory[this.exePointer++] << 8);
                if (this.debug) info("pushconst16", value);
                this.stackPointer -= 4;
                this.writeInt(this.stackPointer, value);
                break;
            }
            case OpCode.PushConst32: {
                let value =
                    this.memory[this.exePointer++] |
                    (this.memory[this.exePointer++] << 8) |
                    (this.memory[this.exePointer++] << 16) |
                    (this.memory[this.exePointer++] << 24);
                if (this.debug) info("pushconst32", value);
                this.stackPointer -= 4;
                this.writeInt(this.stackPointer, value);
                break;
            }
            case OpCode.Dup: {
                if (this.debug) info("dup");
                this.stackPointer -= 4;
                this.writeInt(this.stackPointer, this.readInt(this.stackPointer + 4));
                break;
            }
            case OpCode.Push8: {
                let addr = this.memory[this.exePointer++] | (this.memory[this.exePointer++] << 8);
                if (this.debug) info("push8", addr);
                this.stackPointer -= 4;
                this.writeInt(this.stackPointer, this.memory[addr]);
                break;
            }
            case OpCode.Pop8: {
                let addr = this.memory[this.exePointer++] | (this.memory[this.exePointer++] << 8);
                if (this.debug) info("pop8", addr);
                this.memory[addr] = this.memory[this.stackPointer];
                this.stackPointer += 4;
                break;
            }
            case OpCode.Halt: {
                if (this.debug) info("halt");
                return false;
            }
            case OpCode.Add: {
                if (this.debug) info("add");
                this.writeInt(this.stackPointer + 4, this.readInt(this.stackPointer + 4) + this.readInt(this.stackPointer));
                this.stackPointer += 4;
                break;
            }
            case OpCode.Sub: {
                if (this.debug) info("sub");
                this.writeInt(this.stackPointer + 4, this.readInt(this.stackPointer + 4) - this.readInt(this.stackPointer));
                this.stackPointer += 4;
                break;
            }
            case OpCode.Mul: {
                if (this.debug) info("mul");
                this.writeInt(this.stackPointer + 4, this.readInt(this.stackPointer + 4) * this.readInt(this.stackPointer));
                this.stackPointer += 4;
                break;
            }
            case OpCode.Div: {
                if (this.debug) info("div");
                this.writeInt(this.stackPointer + 4, this.readInt(this.stackPointer + 4) / this.readInt(this.stackPointer));
                this.stackPointer += 4;
                break;
            }
            case OpCode.Mod: {
                if (this.debug) info("mod");
                this.writeInt(this.stackPointer + 4, this.readInt(this.stackPointer + 4) % this.readInt(this.stackPointer));
                this.stackPointer += 4;
                break;
            }
            case OpCode.Inv: {
                if (this.debug) info("inv");
                this.writeInt(this.stackPointer, -this.readInt(this.stackPointer));
                break;
            }
            case OpCode.Abs: {
                if (this.debug) info("abs");
                let val = this.readInt(this.stackPointer);
                this.writeInt(this.stackPointer, val < 0 ? -val : val);
                break;
            }
            case OpCode.Jrnz: {
                if (this.debug) info("jrnz");
                let rel = this.memory[this.exePointer++];
                if (this.readInt(this.stackPointer)) {
                    this.exePointer += rel;
                }
                this.stackPointer += 4;
                break;
            }
            case OpCode.Jrz: {
                if (this.debug) info("jrz");
                let rel = this.memory[this.exePointer++];
                if (!this.readInt(this.stackPointer)) {
                    this.exePointer += rel;
                }
                this.stackPointer += 4;
                break;
            }
            case OpCode.Jr: {
                if (this.debug) info("jr");
                let rel = this.memory[this.exePointer++];
                this.exePointer += rel;
                break;
            }
            case OpCode.Eq: {
                if (this.debug) info("eq");
                this.writeInt(this.stackPointer + 4, this.readInt(this.stackPointer + 4) === this.readInt(this.stackPointer) ? 1 : 0);
                this.stackPointer += 4;
                break;
            }
            case OpCode.Neq: {
                if (this.debug) info("neq");
                this.writeInt(this.stackPointer + 4, this.readInt(this.stackPointer + 4) !== this.readInt(this.stackPointer) ? 1 : 0);
                this.stackPointer += 4;
                break;
            }
            case OpCode.Lt: {
                if (this.debug) info("lt");
                this.writeInt(this.stackPointer + 4, this.readInt(this.stackPointer + 4) < this.readInt(this.stackPointer) ? 1 : 0);
                this.stackPointer += 4;
                break;
            }
            case OpCode.Lte: {
                if (this.debug) info("lte");
                this.writeInt(this.stackPointer + 4, this.readInt(this.stackPointer + 4) <= this.readInt(this.stackPointer) ? 1 : 0);
                this.stackPointer += 4;
                break;
            }
            case OpCode.Bt: {
                if (this.debug) info("bt");
                this.writeInt(this.stackPointer + 4, this.readInt(this.stackPointer + 4) > this.readInt(this.stackPointer) ? 1 : 0);
                this.stackPointer += 4;
                break;
            }
            case OpCode.Bte: {
                if (this.debug) info("bte");
                this.writeInt(this.stackPointer + 4, this.readInt(this.stackPointer + 4) >= this.readInt(this.stackPointer) ? 1 : 0);
                this.stackPointer += 4;
                break;
            }
            case OpCode.Out: {
                info("out", this.readInt(this.stackPointer));
                this.stackPointer += 4;
                break;
            }

            default: {
                throw new Error(`Invalid instruction 0x${op.toString(16)} (${op}) at ${this.exePointer - 1}`);
            }
        }
        return true;
    }
}
